pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -0,0 +1,461 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	csiv1alpha1 "k8s.io/csi-api/pkg/apis/csi/v1alpha1"
+	csiclientset "k8s.io/csi-api/pkg/client/clientset/versioned"
+)
+
+const (
+	// Name of node annotation that contains JSON map of driver names to node
+	annotationKeyNodeID = "csi.volume.kubernetes.io/nodeid"
+)
+
+var nodeKind = v1.SchemeGroupVersion.WithKind("Node")
+
+// nodeInfoManager contains necessary common dependencies to update node info on both
+// the Node and CSINodeInfo objects.
+type nodeInfoManager struct {
+	nodeName      types.NodeName
+	k8s           kubernetes.Interface
+	csiKubeClient csiclientset.Interface
+}
+
+// If no updates is needed, the function must return the same Node object as the input.
+type nodeUpdateFunc func(*v1.Node) (newNode *v1.Node, updated bool, err error)
+
+// Interface implements an interface for managing labels of a node
+type Interface interface {
+	// Record in the cluster the given node information from the CSI driver with the given name.
+	// Concurrent calls to AddNodeInfo() is allowed, but they should not be intertwined with calls
+	// to other methods in this interface.
+	AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error
+
+	// Remove in the cluster node information from the CSI driver with the given name.
+	// Concurrent calls to RemoveNodeInfo() is allowed, but they should not be intertwined with calls
+	// to other methods in this interface.
+	RemoveNodeInfo(driverName string) error
+}
+
+// NewNodeInfoManager initializes nodeInfoManager
+func NewNodeInfoManager(
+	nodeName types.NodeName,
+	kubeClient kubernetes.Interface,
+	csiKubeClient csiclientset.Interface) Interface {
+	return &nodeInfoManager{
+		nodeName:      nodeName,
+		k8s:           kubeClient,
+		csiKubeClient: csiKubeClient,
+	}
+}
+
+// AddNodeInfo updates the node ID annotation in the Node object and CSIDrivers field in the
+// CSINodeInfo object. If the CSINodeInfo object doesn't yet exist, it will be created.
+// If multiple calls to AddNodeInfo() are made in parallel, some calls might receive Node or
+// CSINodeInfo update conflicts, which causes the function to retry the corresponding update.
+func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error {
+	if driverNodeID == "" {
+		return fmt.Errorf("error adding CSI driver node info: driverNodeID must not be empty")
+	}
+
+	if topology == nil {
+		return fmt.Errorf("error adding CSI driver node info: topology must not be nil")
+	}
+
+	err := nim.updateNode(
+		updateNodeIDInNode(driverName, driverNodeID),
+		updateTopologyLabels(topology),
+	)
+	if err != nil {
+		return fmt.Errorf("error updating Node object with CSI driver node info: %v", err)
+	}
+
+	err = nim.updateCSINodeInfo(driverName, driverNodeID, topology)
+	if err != nil {
+		return fmt.Errorf("error updating CSINodeInfo object with CSI driver node info: %v", err)
```

Comment:
Maybe instead of failing, just log a warning for now saying you could not update the object and that topology may not work (caller has to enable CRD in order for it to work). And leave a TODO, once CSINodeInfo object goes beta you can fail instead of log?

Reply:
ACK will update. Currently if an error is returned, the registration handler proceeds to unregister the driver, and we definitely don't wan this if CRD creation is disabled

Reply:
It would have to move to GA right, since a user can still disable the CRD in beta?